### PR TITLE
Support no args gatherers

### PR DESCRIPTION
**Pull request by nelsonkopliku**
This PR relates to https://github.com/trento-project/agent/pull/137, aiming to enable porting of check **438525 /etc/hosts file contains all the cluster nodes** in the new engine.

<details>
  <summary>See previous considerations for more context</summary>
  <br />
I am attempting to add the Check Specification for **438525: /etc/hosts file contains all the cluster nodes**

Here's the current implementation:
```sh
for i in $(crm_node -l | cut -d' ' -f2); do
  grep -v '^[[:space:]]*#' /etc/hosts | egrep -q "[[:space:]]+${i}(\.|[[:space:]]+|$)" || exit 1
done
exit 0
```

Meaning that
given a list of node names (`crm_node -l | cut -d' ' -f2`), those node names must be present in `/etc/hosts`

I am trying to leverage the recently added `hosts` gatherer, but I am having difficulties with how it currently works, there's something that does not click.
```yaml
facts:
  - name: hosts_node1
    gatherer: hosts
    argument: node1
  
  - name: hosts_node2
    gatherer: hosts
    argument: node2
```

Given the check specifies that  **/etc/hosts file contains all the cluster nodes** I believe that:
1. We are probably missing a gatherer that returns node names of a cluster (`$(crm_node -l | cut -d' ' -f2)`)
2. the current hosts gatherer itself might need to be revisited, or new ones added

How could this Check Specification look like?
```yaml
facts:
  - name: configured_cluster_nodes
    gatherer: cluster_nodes
    # argument: ?? # might not be needed

  - name: available_host_names
    gatherer: hosts
    # argument: ?? # might not be needed

expectations:
  - name: all_cluster_nodes_are_available_in_hosts_configuration
    expect: facts.configured_cluster_nodes.all(|cluster_node| cluster_node in facts.available_host_names)
```
Expectation expression might differ based on what the gatherers return.

Maybe an offline discussion might help.
</details>

Mainly introduces the possibility to have gatherers that don't need arguments.
```yaml
facts:
  - name: configured_cluster_nodes
    gatherer: cluster_nodes # THIS GATHERER IS NOT AVAILABLE, YET

  - name: available_host_names
    gatherer: hosts

expectations:
  - name: all_cluster_nodes_are_available_in_hosts_configuration
    expect: facts.configured_cluster_nodes.all(|cluster_node| cluster_node in facts.available_host_names.value)
```
`cluster_nodes` is a possible gatherer that returns the list of nodes of a cluster (`crm_node -l | cut -d' ' -f2`).
Alternatively the list of node could be provided via the `env`.
Anyway something needs to be implemented, be it an extra gatherer or providing that list when an execution is started.

```elixir
context = %{
  "facts" => %{
    "configured_cluster_nodes" => [
      "node1",
      "node2"
    ],
  }
}

# or

context = %{
  "env" => %{
    "configured_cluster_nodes" => [
      "node1",
      "node2"
    ],
  }
}
```

while `hosts` is the updated version here https://github.com/trento-project/agent/pull/137 which could potentially return all the entries in the `/etc/hosts`, without any argument provided.

```elixir
context = %{
  "facts" => %{
    "available_host_names" => %{
      "value" => %{
        "node1" => %{
          "value" => [
            "123.456.789",
            "1.1.1.1"
          ]
        }
      }
    }
  }
}
```

These things together could allow us to specify the Check under consideration _438525: /etc/hosts file contains all the cluster nodes_ in one of these two ways

```rust
facts.configured_cluster_nodes.all(|cluster_node| cluster_node in facts.available_host_names.value)
# or
env.configured_cluster_nodes.all(|cluster_node| cluster_node in facts.available_host_names.value)
```

Keeping a bit on hold for now, switching to simpler Checks, so that we can have extra thoughts from the team about this and possibly more complex ones and iterate.

Open points are:
- ~gatherer's `argument` nullability at contracts level (it should be already supported)~
- where to get the cluster nodes from? A gatherer? From the execution environment?
- returned facts structure (**facts.available_host_names**`.value`)

### CODE REVIEW

**Comment by rtorrero** [comment]
Hey there @nelsonkopliku, I mostly agree with point `1` but I'm not sure what do you mean by "the current hosts gatherer itself might need to be revisited, or new ones added", what would you change about it?

In general lines, now that we have not constrained by the limitations of abacus, I'd go for reusable, generic gatherers and rely on the expression evaluation to combine different gatherers to make more complex checks. In other words, what I'd like to avoid is having gatherers so specific (i.e. "cluster_nodes_ip_from_hosts" gatherer) that will only be possible to use in a few cases.

**Review by arbulu89 — COMMENTED** [review]
I don't think that making this field nullable is a good idea. I would only do this is the meaning of a null value and empty string is different.
The `argument` is only used in the agent (so here we only deliver it there), and in golang, we would be required to make the string a pointer, with all the issues that this brings (type sanity). And besides that, we would need to change the contract itself.

I would vote of using an empty string when the argument is not needed. In the agent side it would be as simple as comparing an empty string.

Most probably, if we want to implement the multi argument gatherers, this must be changed, so I would go with the easiest solution by now.

File: priv/catalog/438525.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+id: "438525"
```
[the remark this answers is not in the document]
Comment (fabriziosestito) [reply priv/catalog/438525.yaml:1]: I agree, let's please keep this out.

File: test/fixtures/catalog/argumentless_gatherer.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+id: argumentless_gatherer
```
[the remark this answers is not in the document]
Comment (fabriziosestito) [reply test/fixtures/catalog/argumentless_gatherer.yaml:1]: I agree. We could also create just one test check in the fixtures and refactor the tests a bit, I will take care of this in a follow up PR if it's ok for you guys.

Also, I find the name "argumentless" super confusing. 
I found this commonly referred to as "no arguments" or "no args" basically in every codebase.

File: test/messaging/mapper_test.exs [hunk 3]
```diff
@@ -15,73 +15,111 @@ defmodule Wanda.Messaging.MapperTest do
...
+                       gatherer: "gatherer_3",
+                       name: "fact_3"
+                     }
+                   ]
+                 }
+               ]
+             } = Mapper.to_facts_gathering_requested(execution_id, group_id, targets, checks)
+    end
+
+    test "should map to a FactGatheringRequested event, with argument-less gatherers" do
```
Comment (fabriziosestito) [inline test/messaging/mapper_test.exs:88]: couldn't we smash this with the previous test?

File: lib/wanda/catalog.ex [hunk 4]
```diff
@@ -100,6 +100,14 @@ defmodule Wanda.Catalog do
     }
   end
 
+  defp map_fact(%{"name" => name, "gatherer" => gatherer}) do
```
Comment (fabriziosestito) [inline lib/wanda/catalog.ex:103]: Based on @arbulu89 comments about the empty string we could simply:

```elixir
defp map_fact(%{"name" => name, "gatherer" => gatherer} = fact) do
 %Fact{
      name: name,
      gatherer: gatherer,
      argument: Map.get(fact, "argument", "")
    }
end
```

**Review by fabriziosestito — CHANGES_REQUESTED** [review]
Overall I like the idea of supporting empty arguments.
Since ATM we just have string type, empty strings are easier to manage on the agent side and I would vote to go in that direction, with the promise to revise this if we will discover that the gatherers could have other types as well. 

Said so, I left some comments since I don't think we need to put so much focus on a default case, so the tests could be simplified/slimmed down.

File: priv/catalog/438525.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (nelsonkopliku) [reply priv/catalog/438525.yaml]: Yup

**Comment by nelsonkopliku** [comment]
Thanks @fabriziosestito and @arbulu89 for the reviews! Very useful! :rocket:

**Review by fabriziosestito — APPROVED** [review]
LGTM! Could please track the debt about refactoring catalog test fixtures in Wanda?

**Pull request merged**